Overhaul transaction request logic

sipa opened the pull request:
This replaces the transaction request logic with an encapsulated class that maintains all the state surrounding it. By keeping it stand alone, it can be easily tested (using included unit tests and fuzz tests).

The new logic is close in spirit to the current one, but much easier to specify I think:
* Always prefer outbound peers over inbound peers, as long as any are available.
* Don't download from inbound peers during the first 2 s after announcement, to give time for announcements from outbound peers to come in.
* The peer who is the very first to announce a txid is preferred (compared to other inbounds or other outbounds), so that chains of dependent transactions announced by one peer will generally not be split out to other peers.
* After the peer that announces a txid first, requests go to uniformly selected peers (outbounds first; inbounds that are past their 2 s delay afterwards), to minimize the impact an attacker can have (as attackers can easily race the network and be first).
* A txid is never requested twice from the same peer, as long as other candidates remain (even when reannounced).
* Transactions are requested from new candidates as soon as old requests expire, or NOTFOUND is received, or invalid-witness transactions are received (i.e. this includes the functionality added by, and replaces, #18238).

One significant change is the removal of the in-flight limit of 100 txids. I think this limit is not exactly desirable (e.g. when a txid is only announced by one peer, we should request it, regardless of how many other txids we're already waiting for), and with the changes to handle NOTFOUND immediately, there is less problems with high limits. This implementation's performance is also only determined by the number of total transaction/peer pairs tracked (O(log n)), not by the number in-flight, removing another reason for having that limit.

A few further improvements could be investigated and relatively easily added if desirable:
* Biasing away from peers with lots of in-flight requests (but not outright limiting like now).
* Making the 2s inbound delay Poisson-distributed.
* Adding a wtxid flag to entries to make it compatible with #18044.
* Adding a time limit for the time between announcement and request.

While the implementation is non-trivial due to performance and resource usage considerations, it has a significantly simpler exact specification. This specification is documented in src/txrequest.h, and reimplemented naively in the src/test/fuzz/txrequest.cpp fuzz test. It may be a good idea to review those first to understand the desired semantics before digging into the optimized implementation.

sipa committed:
Add txrequest module

sipa committed:
Add txrequest unit tests

sipa committed:
Switch net_processing over to txrequest-based fetching

sipa committed:
More accurate removal of transactions from g_txrequest

sipa committed:
Delete limitedmap as it is unused

sipa committed:
Add txrequest fuzzer that compares with naive implementation

sipa commented:
> Why invalid transactions are re-requested? Maybe I can figure out after reading the code, but right now this is confusing.

It requires invalid witnesses. When you request a txid T from peer A, and get a response with a transaction with invalid witness, but txid T, then the announcement (T,A) gets marked as COMPLETED.

amitiuttarwar commented:
Concept ACK

jnewbery commented on src/txrequest.h:
```diff
@@ -0,0 +1,342 @@
+    const uint64_t m_k0, m_k1;
+public:
+    explicit PriorityComputer(bool deterministic);
+    PriorityComputer(const PriorityComputer& other) = default;
```
This isn't required. We never copy construct a priority computer.

sipa replied:
Fixed.

jnewbery commented on src/txrequest.h:
```diff
@@ -0,0 +1,342 @@
+ * - ReceivedResponse(peer, txid) converts any CANDIDATE or REQUESTED entry to a COMPLETED one. It should be called
+ *   whenever a transaction or NOTFOUND was received from a peer. When the transaction is acceptable, AlreadyHaveTx
+ *   should be called instead of (or in addition to) this operation.
+
```
missing `*`. I don't know if this messes with doxygen parsing.

sipa replied:
Fixed.

jnewbery commented on src/txrequest.h:
```diff
@@ -0,0 +1,342 @@
+
+    //! The various states a (txid,node) pair can be in.
+    //! Note that CANDIDATE is split up into 3 substates (NEW, BEST, OTHER), allowing more efficient implementation.
+    enum class State : uint8_t {
```
Is there a reason to specify the underlying type here? Is it required so that it can be packed into 3 bits in the Entry bitfield below?

sipa replied:
Yeah, my concern is that otherwise an unsigned type may be picked, and I don't know how that would interact with it being used as a bitfield.

jnewbery replied:
I'm getting this build warning:

```
In file included from txrequest.cpp:5:0:
./txrequest.h:158:25: warning: ‘TxRequestTracker::Entry::m_state’ is too small to hold all values of ‘enum class TxRequestTracker::State’
         State m_state : 3;
                         ^
```

configure output:

```
  target os     = linux
  build os      = linux-gnu

  CC            = /usr/bin/ccache gcc
  CFLAGS        = -g -O2
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = /usr/bin/ccache g++ -std=c++11
  CXXFLAGS      =   -fstack-reuse=none -Wstack-protector -fstack-protector-all  -Wall -Wextra -Wformat -Wvla -Wswitch -Wformat-security -Wredundant-decls -Wunused-variable -Wdate-time -Wsign-compare  -Wno-unused-parameter -Wno-implicit-fallthrough   -g -O2 -fno-extended-identifiers
  LDFLAGS       = -pthread  -Wl,-z,relro -Wl,-z,now -pie  
  ARFLAGS       = cr
```

gcc version:

```
→ gcc --version
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
```

ajtowns replied:
Looks like that warning is kind-of a bug in gcc (you can apparently put any value from the underlying type in the enum, and it's complaining that they won't all fit), but it's not fixed until gcc 8.4 or 9.3 -- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61414

sipa replied:
@jnewbery Does it help if you remove the `: uint8_t`?

If this is too annoying we can just store the state as a non-`enum` type.

sipa replied:
Done. Changed to store it as a `uint8_t`.

jnewbery commented on src/txrequest.h:
```diff
@@ -0,0 +1,342 @@
+        //! A REQUESTED entry.
+        REQUESTED,
+        //! A CANDIDATE entry that's not CANDIDATE_NEW or CANDIDATE_BEST.
+        CANDIDATE_OTHER,
```
nit: sort this above `REQUESTED` so it's next to the other `CANDIDATE_` entries.

sipa replied:
That doesn't work directly, as EntryTxid's sort order relies on the ordering of these fields. It could use its own enum with the necessary order, that ExtractTxid maps to, but that feels like overkill.

ajtowns replied:
Maybe worth adding a comment in the enum definition that the ordering is relied upon by the code that uses the ByTxid indexes. Reordering in ExtractTxid seems like overkill to me too fwiw.

sipa replied:
Done.

jnewbery replied:
Agree that a comment is sufficient.

jnewbery reviewed (commented):
Concept ACK

There's a lot to review here. It'll take a while to work through it.

I've left a few small style comments from a quick read through the new header file.

ariard commented:
Concept ACK, I think that's a great move to dissociate message paths by class (addr, block, transaction, filters) instead of having all of them melt down in `net_processing`.

I will do an approach ACK soon, but sounds to me this add more reliance on `boost` ? Also as it's a critical part of the codebase how do we build confidence of the correctness of this new transaction request logic implementation with regards to substituted old part ? (It would be great to come with a model for this as I think it would be desirable to do the same kind of changes for the messy `COMPACT_BLOCKS part of the codebase)

ajtowns commented:
> sounds to me this add more reliance on `boost` ?

It uses boost multi_index, which is also needed by the mempool (which does lookups by txid and feerate eg). If we found (or created) a replacement for that there, it should work here too, so I don't think it's much of an increase in the dependency.

sipa commented:
@ariard As @ajtowns said, we're already relying on multi_index for the mempool. I also spent a week or two thinking about how to implement this with custom data structures, but it's... really hard to match the same performance/memory usage characteristics without multi_index. So I think it's just the right tool for the job, plus it's a headers-only library (so it doesn't introduce runtime dependencies, which is a large part of what makes boost annoying).

As far as correctness/desirability of the behavior goes:

* Reasoning about whether the behavior is desirable, when considered standalone, I believe is much easier than the current one. The current behavior is very much defined by interactions of randomized insertion orders in a queue, and has few guarantees about what those actually accomplish. You can look at the fuzz test, which contains a complete reimplementation of the semantics (without optimized performance).

* At a (very) high level you can see that the existing behavior is maintained (apart from the max-in-flight rule) by looking at the changes in test/functional/p2p_tx_download.py. But at a lower level it's again hard to say much because the existing behavior is already so hard to describe exactly.

hebasto commented on src/txrequest.h:
```diff
@@ -0,0 +1,332 @@
+    std::chrono::microseconds GetTimeout() const { return m_timeout; }
+};
+
+#endif
```
3d5b4f20015246b7a0dd74b2dc89f389204dd9b8

```suggestion
#endif // BITCOIN_TXREQUEST_H
```

See: [Source code organization](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization)

sipa replied:
Done.

hebasto commented on src/txrequest.h:
```diff
@@ -0,0 +1,332 @@
+ * - REQUESTED (txid, peer, exptime) entries representing txids that have been requested, and we're awaiting a
+ *   response for from that peer. The exptime value determines when the request times out.
+ *
+ * - COMPLETED (txid, peer) entries representing txids that have been requested from a peer, but they timed out, a
+ *   NOTFOUND message was received for them, or an invalid response was received. They're only kept around to prevent
+ *   downloading them again. If only COMPLETED entries exist for a given txid remain (so no CANDIDATE or REQUESTED
+ *   ones), all of them are deleted (this is an invariant, and maintained by all operations below).
```
3d5b4f20015246b7a0dd74b2dc89f389204dd9b8

nanonit: the word "completed" associates with a "success" for me (but I'm not a native English speaker though). These entries are not processed successfully, i.e., a transaction has not been received. Maybe rename `COMPLETED` to more general `PROCESSED` or something similar?

sipa replied:
I've struggled with a good name for this, actually, and changed it a few times. The problem is that it corresponds to a number of scenarios:

* We've received the transaction (but we don't know (yet) if it's a valid one)
* We've received the transaction with an invalid witness
* We've received a NOTFOUND message for the transaction
* The request timed out

I don't think `PROCESSED` is an improvement, but I admit `COMPLETED` doesn't perfect either.

hebasto commented on src/txrequest.h:
```diff
@@ -0,0 +1,332 @@
+ * - AlreadyHaveTx(txid) deletes all entries for a given txid. It should be called when a transaction is successfully
+ *   added to the mempool, seen in a block, or for whatever reason we no longer care about it.
+ *
+ * - ReceivedResponse(peer, txid) converts any CANDIDATE or REQUESTED entry to a COMPLETED one. It should be called
+ *   whenever a transaction or NOTFOUND was received from a peer. When the transaction is acceptable, AlreadyHaveTx
+ *   should be called instead of (or in addition to) this operation.
```
3d5b4f20015246b7a0dd74b2dc89f389204dd9b8

It is not clear for me why the `ReceivedResponse()` "should be called whenever a transaction ... was received from a peer". It is not the case for a `COMPLETED` entry, no?

sipa replied:
`ReceivedResponse` is what marks the entry as `COMPLETED`.

hebasto reviewed (commented):
Concept ACK. The interface presented in the `txrequest.h` header is clean and simple enough.

jnewbery commented on src/txrequest.h:
```diff
@@ -0,0 +1,341 @@
+        bool IsSelectable() const { return m_state == State::CANDIDATE_OTHER || m_state == State::CANDIDATE_BEST; }
+
+        //! Construct a new entry from scratch
+        Entry(const uint256& txid, uint64_t peer, bool inbound, State state, std::chrono::microseconds time, uint64_t
```
I'm trying not to leave too many nits at this stage, but what kind of maniac breaks lines between the type and parameter name?!

sipa replied:
🤪

sipa replied:
Fixed.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1951,6 +1835,7 @@ void static ProcessOrphanTx(CConnman* connman, CTxMemPool& mempool, std::set<uin
         if (setMisbehaving.count(fromPeer)) continue;
         if (AcceptToMemoryPool(mempool, orphan_state, porphanTx, &removed_txn, false /* bypass_limits */, 0 /* nAbsurdFee */)) {
             LogPrint(BCLog::MEMPOOL, "   accepted orphan tx %s\n", orphanHash.ToString());
+            g_txrequest.AlreadyHaveTx(orphanHash);
```
I don't think this will ever do anything. By the time we process this orphan transaction, we will have already called `AlreadyHaveTx()` in a `SendMessages()` call.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2817,16 +2702,14 @@ void ProcessMessage(
         if (!AlreadyHave(inv, mempool) &&
             AcceptToMemoryPool(mempool, state, ptx, &lRemovedTxn, false /* bypass_limits */, 0 /* nAbsurdFee */)) {
             mempool.check(&::ChainstateActive().CoinsTip());
+            g_txrequest.AlreadyHaveTx(inv.hash);
```
There's a slight inconsistency here: if the transaction is successfully accepted to the mempool, then we call `AlreadyHaveTx()` immediately to remove it from `g_txrequest`, but if it's an orphan, then we only call `AlreadyHaveTx()` the next time a peer would potentially request it in `SendMessages()` (and I think the `AlreadyHaveTx()` call in `ProcessOrphanTx()` is basically always a no-op).

For consistency, perhaps we should only call `AlreadyHaveTx()` in `SendMessages()`.

ajtowns replied:
Thinking about the case where someone malleates the witness for a tx relayed by txid not wtxid; should we maybe ... I don't know, track the other peers that offered the tx as part of the orphan record, so that we can try getting the correct witness data once we have the parents and can actually see if maybe the witness data was malleated?

jnewbery replied:
Pieter and I were discussing just this yesterday. The behaviour you describe (not rerequesting an orphan transaction from another announcing peer if we later discover that its witness is mutated) also exists in master. As soon as we put a transaction in `mapOrphanTransactions` it'll return true from `AlreadyHave()` and will then be removed from each peer's `TxDownloadState`.

I think it's ok to keep this behaviour. Orphan handling is to some extent best effort, so shouldn't be relied on for timely transaction propagation. If an adversary is able to poison a transaction in the orphan set, then we won't rerequest it from other peers when `AcceptToMemoryPool()` fails, but it won't be added to our recent rejects filter, and if just one other peer later announces it to us, we'll attempt to fetch it again.

sipa replied:
I prefer to keep the fast-path AlreadyHaveTx calls when transactions/blocks come in, as they may in some cases be much faster than the AlreadyHave detection after GetRequestable (in particular, when the only alternative peer(s) are timing out).

You're right though that the code is inconsistent: orphans in the orphan pool are considered AlreadyHave, so txids should be reported as AlreadyHaveTx as soon as they're added to the orphan pool. I've adopted that approach now.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,341 @@
+    public:
+        EntryTxidExtractor(const PriorityComputer& computer) : m_computer(computer) {}
+        using result_type = EntryTxid;
+        result_type operator()(const Entry& entry) const { return entry.ExtractTxid(m_computer); }
```
Yes sorry John, you're obviously an idiot if you don't know the boost reference documentation by heart ;)

Done.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,341 @@
+    struct Entry {
+        //! Txid that was announced.
+        const uint256 m_txid;
+        //! For CANDIDATE{OTHER,_DELAYED,_BEST} the reqtime; for REQUESTED the exptime
```
Done.

hebasto commented on src/txrequest.h:
```diff
@@ -0,0 +1,342 @@
+    {
+        uint64_t low_bits = 0;
+        if (!first) {
+            low_bits = CSipHasher(m_k0, m_k1).Write(txid.begin(), txid.size()).Write(peer).Finalize() >> 1;
```
272a615fad6f93e1b7f9b6245ce9090d07fb4a51

There is `SipHashUint256Extra(uint64_t, uint64_t, uint256&, uint32_t)` in the `crypto/siphash` module.
Is it worth to add the reusable `SipHashUint256Extra64(uint64_t, uint64_t, uint256&, uint64_t)` there?

sipa replied:
I think this may make sense as a follow-up, but probably deserves benchmarks etc.

elichai replied:
Somewhat related ;) https://github.com/bitcoin/bitcoin/pull/18014

hebasto commented on src/txrequest.h:
```diff
@@ -0,0 +1,342 @@
+        //! Whether it's from an outbound peer or not.
+        const bool m_inbound : 1;
+        //! What state this announcement is in
+        //! This is a uint8_t instead of a State to silence a GCC warning.
+        uint8_t m_state : 3;
```
272a615fad6f93e1b7f9b6245ce9090d07fb4a51

No warnings are observed on GCC 9.3.0.
What GCC version you are referring to?

sipa replied:
See https://github.com/bitcoin/bitcoin/pull/19184#discussion_r444545289

hebasto commented on src/txrequest.cpp:
```diff
@@ -0,0 +1,241 @@
+std::vector<uint256> TxRequestTracker::GetRequestable(uint64_t peer, std::chrono::microseconds now)
+{
+    // Move time.
+    SetTimePoint(now);
```
272a615fad6f93e1b7f9b6245ce9090d07fb4a51

Could passing the `peer` to `SetTimePoint()` improves efficiency by not processing entries related to other peers?

sipa replied:
That's a good suggestion, but I don't think it would actually be desirable (but fortunately, it also wouldn't gain any efficiency). An earlier design of this code worked like that, but I decided to move off that.

Say we're calling `GetRequestable` for peer A. There is a txid T that was announced by peer A and peer B. For some reason, peer B is currently the `CANDIDATE_BEST` peer for T, but its exptime is in the past. That means that the (T,B) pair in reality should be `COMPLETED` now (but the data structure hasn't been updated to reflect that), and peer A is actually the best match. If `GetRequestable` for A would only invoke `SetTimePoint` for currently-`CANDIDATE_BEST` entries, T won't be returned, at it'll only be when B calls `GetRequestable` that its expiration happens. This is probably not terrible behavior, but makes it *much* harder to exactly specify the behavior, and perhaps it introduces a bias in the selection - it's hard to say.

Thankfully, it also doesn't gain anything. Every (txid,peer) pair only ever transitions once to `COMPLETED`, so certainly in bulk work there cannot be any difference. Furthermore, `SetTimePoint` only iterates over all expired entries in its first loop, as they're sorted by time. So no time is wasted processing unexpired entries.

A similar reasoning applies to the second loop.

hebasto commented on src/txrequest.h:
```diff
@@ -0,0 +1,342 @@
+        REQUESTED,
+        //! A CANDIDATE entry that's not CANDIDATE_NEW or CANDIDATE_BEST.
+        CANDIDATE_OTHER,
+        //! A COMPLETED entry.
+        COMPLETED,
```
IIUC, `TxRequestTracker::MakeCompleted()` relies on the fact that `COMPLETED` is the last entry in the enum `State`.
Could a relevant comment be add to preserve any breakdown in the future?

sipa commented:
@gmaxwell suggested that the "first" feature (the very first peer to announce a transaction gets priority over other outbound or other inbounds) is changed to instead apply to the first not-max-in-flight peer. The idea is that this would prevent an attacker who races announcements, but doesn't respond, will (as soon as they reach their max-in-flight) not be able to interfere with chains of dependent transactions (which we'd like to request from the same peer).

This seems like a good idea, and easy to implement, but it'll require some changes to tests.

sipa commented on src/txrequest.cpp:
```diff
@@ -0,0 +1,241 @@
+    while (it != index.end() && it->m_peer == peer) {
+        // Check what to continue with after this iteration. Note that 'it' may change position, and std::next(it)
+        // may be deleted in the process, so this needs to be decided beforehand.
+        auto it_next = (std::next(it) == index.end() || std::next(it)->m_peer != peer) ? index.end() : std::next(it);
```
Yes, iterator validation for multi_index entries is like std::set; erasing entries does not affect the validity of other entries.

practicalswift commented:
Concept ACK

Very nice fuzzing harness!

sipa commented:
I've made a few changes:
* Instead of having an outbound/inbound flag, `ReceivedInv` now just takes an explicit delay parameter. Entries with delay 0 are preferred. This matches reality more closely (as the `outbound` property wasn't exactly for outbound peers, but also other whitelisted ones).
* Renamed CANDIDATE_NEW to CANDIDATE_DELAYED
* Renamed CANDIDATE_OTHER to CANDIDATE_READY

As there are still other changes I want to make to the behavior, I've marked this as Draft for now.

ariard commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+    //! For whatever reason, we no longer need this txid. Delete any data related to it.
+    void AlreadyHaveTx(const uint256& txid);
+
+    //! We received a new inv, enter it into the data structure.
```
259fb92

We call `ReceivedInv` on the behalf of a peer sending us an orphan tx. Is following scenario plausible ?
* Malicia sends N txn to Alice and N' conflicting to Bob
* Malicia sends txB child spending from all parent of N to Alice
* Alice sends txB to Bob, at reception Bob calls `RequestTx/ReceivedInv` on behalf of Alice
* Malicia repeats this trick until reaching Alice's  MAX_PEER_TX_ANNOUNCEMENTS (and also avoid hitting MAX_STANDARD_TX_WEIGHT for orphan)
* When Bob calls `GetRequestable` at GETDATA sending to Alice, only MAX_GETDATA_SZ will be effectively processed, the rest obstructing Alice's announced invs until GETDATA_TX_INTERVAL expiration
* Alice sends transaction for the MAX_GETDATA_SZ set
* Malicia may repeat timely repeat trick to maintain transaction propagation blocking anticipating Alice-Bob round-trips

Attack might seem costly because you have to fulfill Alice mempool with parent transaction but making them conflicted with the rest of network mempools avoid to effectively pay their feerate.

sipa replied:
I think you're confused. `MAX_GETDATA_SZ` doesn't stop processing; it only makes the request be split up into multiple GETDATA messages.

ariard replied:
You're right on `MAX_GETDATA_SZ`.

Still, do you think that a third-party can delay transaction announcement between Alice and Bob by forcing her to send him orphans until reaching MAX_PEER_TX_ANNOUNCEMENTS ?

sipa replied:
I need to think about this more.

ariard commented on src/txrequest.cpp:
```diff
@@ -0,0 +1,238 @@
+        return false;
+    }
+
+    // Mark the entry COMPLETED, and select the best best entry if needed.
```
259fb92

nit: The first-sorted _READY ?

sipa replied:
Done.

ariard commented on src/txrequest.cpp:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    while (!m_index.empty()) {
+        // If time went backwards, we may need to demote CANDIDATE_BEST and CANDIDATE_READY entries back
```
Time going backward wouldn't be a hint of some clock manipulation and thus we should interpret this as a system failure and halting ? AFAICT, `GetAdjustedTime()` can go backward due to relying on peers clocks but we are relying on `GetTime<T>` here.

sipa replied:
Time jumping backwards can happen for completely innocuous reasons, like leap seconds. Only monotonic clocks are guaranteed not to do that.

ariard commented on src/txrequest.cpp:
```diff
@@ -0,0 +1,238 @@
+
+void TxRequestTracker::AlreadyHaveTx(const uint256& txid)
+{
+    auto it = m_index.get<ByTxid>().lower_bound(EntryTxid{txid, State::CANDIDATE_DELAYED, 0});
```
259fb92

Maybe add as a comment "If transaction has already been evaluated once for this block tip, delete unconditionally all its entries."

Is this a slight behavior change ?  Previously we would have only delete transaction from download for this peer. If you received a block between peers processing you have flushed `recentRejects` and so here return a different evaluation.

sipa replied:
> Maybe add as a comment "If transaction has already been evaluated once for this block tip, delete unconditionally all its entries."

I don't understand what reevaluation has to do with that. And `TxRequestTracker` doesn't know or care about the concept of blocks. It's just told that a transaction is no longer needed, so it can forget announcements about it. Why the caller decided that was the right thing to do is not its concern.

> Is this a slight behavior change ?

The whole thing is a giant behavior change - it's changing out the semantics for something conceptually similar, but very different in the details.

ariard replied:
> Why the caller decided that was the right thing to do is not its concern.

Right, I understand comments should be concerned with other transaction relay mechanisms.

ariard commented on src/txrequest.cpp:
```diff
@@ -0,0 +1,238 @@
+
+void TxRequestTracker::ReceivedResponse(uint64_t peer, const uint256& txid)
+{
+    // We need to search the ByPeer index for both (peer, false, txid) and (peer, true, txid).
```
259fb92

The ByPeer index second member as true signals a state equal to CANDIDATE_BEST but I'm not sure if you can effectively hit here. An entry state is _BEST only between `GetRequestable` and either `RequestedTx/AlreadyHaveTx`, both of them respectively transitioning to REQUESTED/entry removal.

sipa replied:
As explained elsewhere, no, BEST exists for much longer than that, and I'm 99% sure this can be hit.

ariard commented on src/txrequest.cpp:
```diff
@@ -0,0 +1,238 @@
+    // Nothing to be done if it's already COMPLETED.
+    if (it->GetState() == State::COMPLETED) return true;
+
+    if ((it == m_index.get<ByTxid>().begin() || std::prev(it)->m_txid != it->m_txid) &&
```
259fb92

nit: I think this control flow conditional could be replaced by a method `IsLastCompleted` like we have `IsSelected/IsSelectable`. Easier to reason.

sipa replied:
Good idea, done. Added `IsOnlyNonCompleted`.

ariard commented on src/net_processing.cpp:
```diff
@@ -2599,7 +2482,7 @@ void ProcessMessage(
                     return;
                 } else if (!fAlreadyHave && !chainman.ActiveChainstate().IsInitialBlockDownload()) {
-                    RequestTx(State(pfrom.GetId()), inv.hash, current_time);
+                    RequestTx(pfrom, inv.hash, current_time);
```
4f4f2f2

If `fAlreadyHave==true` call `AlreadyHaveTx` to clean transaction entries early ?

sipa replied:
I don't think that's necessary - we just have to call `ForgetTx` whenever a transaction *becomes* alreadyhave, and prevent adding announcements for alreadyhave ones. If that's done correctly, an addition `ForgetTx` here would always be a no-op.

ariard commented on src/net_processing.cpp:
```diff
@@ -4349,59 +4221,19 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
-                    tx_process_time.emplace(next_process_time, txid);
+                LogPrint(BCLog::NET, "Requesting %s peer=%d\n", inv.ToString(), pto->GetId());
+                vGetData.push_back(inv);
+                if (vGetData.size() >= MAX_GETDATA_SZ) {
```
4f4f2f2

Maybe you can parameterize `GetRequestable` with MAX_GETDATA_SZ to avoid leftover marked as REQUESTED but not effectively fetched and thus wasting a `GETDATA_TX_INTERVAL` for them ?

sipa replied:
GetRequestable could be parametrized with a limit for the number of returned responses, but that would just be to avoid having it return the same thing multiple times. I don't understand what you mean with "avoid leftover marked as REQUESTED" and "wasting a `GETDATA_TX_INTERVAL`.

ariard replied:
This comment was wrong and part of my misunderstanding you pointed above.

ariard commented on src/test/txrequest_tests.cpp:
```diff
@@ -0,0 +1,237 @@
+    CheckRequestable(tracker, peers[1], now, {txid2, txid1}); // even when the clock went backwards during announcements.
+}
+
+void TestAll()
```
75fd5dc

I think `AlreadyHaveTx` and time-goes-backward of `SetTimePoint` aren't covered? I'll try to add them soon, also `ReceivedResponse` should be called unconditionally at the end of each test to verify clean state ?

sipa replied:
The unit tests are completely rewritten. There aren't any for the time-moving-backward though, I'll try to add that soon.

sipa replied:
Actually, that's not necessary anymore. As `ReceivedInv` doesn't have a "now" argument anymore, the time of announcement clearly doesn't matter, simply by the API being agnostic of it. I'm adding a test that has a reqtime that decreases, though.

ariard reviewed (commented):
Approach ACK, mostly asking for specification/behavior clarifications. I see better how boost::multi_index_container fits well for the job here, not a concern to me anymore.

Overall, great work!

sipa commented:
@ariard Thanks for the thorough review. I'll go over your comments soon, but know there are a few imminent changes first (namely the suggestion https://github.com/bitcoin/bitcoin/pull/19184#issuecomment-652185602, and adding support for wtxid relay tracking from #18044). If you want to help working on the tests, it's probably best to wait until those are complete.

jonatack commented:
Concept ACK, on first read-through the changes look good, will review when this comes out of draft status.

instagibbs commented:
concept ACK

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+ * - Entries are called delayed if they have a reqtime different from std::chrono::microseconds::min(). This is the
+ *   case for entries from inbound non-whitelisted nodes. The priority for delayed entries is always higher than
+ *   non-delayed ones, so non-delayed ones are always preferred.
+ * - Within the set of delayed peers and within the set of non-delayed peers:
```
The "first" concept works differently now. Have a look at https://github.com/bitcoin/bitcoin/pull/19988 to see if your comment still applies.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+ *
+ * === High level behavior ===
+ *
+ * Transaction downloads are attempted in random order, first choosing among non-delayed peers, then delayed
```
See #19988. The "delayed"/"nondelayed" distinction doesn't exist anymore. There is an explicit separate boolean to `ReceivedInv` now for preferred/nonpreferred, and a way to specify reqtime explicitly. (in particular, because txid peers get a delay, even though they're not preferred like outbounds are - so these concepts needed to be split).

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+ * ones whose delay has passed. When a NOTFOUND is received, an invalid transaction is received, or a download
+ * times out (after a configurable delay), the next candidate is immediately scheduled according to the same rules
+ * above. The same transaction is never re-requested from the same peer, unless the transaction was forgotten about
+ * in the mean time. This happens whenever no candidates remain, or when a transaction is successfully received.
```
It is just clarifying when a transaction is forgotten about, which can happen when either no candidates for a tx remain, or when the transaction is successfully received. Only the first of those two matters to re-requesting, but it's still important to specify that transactions are forgotten about when they're successfully received. See if it's clearer now.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+ * - ReceivedInv(txid, peer) adds a new CANDIDATE entry, unless one already exists for that (txid, peer) combination
+ *   (whether it's CANDIDATE, REQUESTED, or COMPLETED).
+ *
+ * - DeletedPeer(peer) deletes all entries for a given peer. It should be called when a peer goes offline.
```
No, the description here is about the specification, where the distinction between the various CANDIDATE classes doesn't exist; that's an implementation detail.

There are many caveats to having in-flight limits... I couldn't come up with a way to do it basically, and then realized it's not desirable either (if a peer is overloaded, but still the only one to announce a tx, we still want to fetch it!). There are many ways by which overfull peers could be deprioritized safely though without introducing a hard cap.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+ *   non-const operation removes the ability to call RequestedTx.
+ */
+class TxRequestTracker {
+    //! Configuration option: timeout after requesting a download.
```
It was made configurable to simplify testing.

In the new PR the setting is gone, and `RequestedTx` now takes an explicit `exptime` parameter instead.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+        CANDIDATE_DELAYED,
+        //! The best CANDIDATE for a given txid; only if there is no REQUESTED entry already for that txid.
+        //! The CANDIDATE_BEST is the lowest-priority entry among all CANDIDATE_READY (and _BEST) ones for that txid.
+        CANDIDATE_BEST,
```
It's not that simple, unfortunately.

`GetRequestable` moves time (making things best/ready if they pass reqtime, making them completed if they pass exptime) for _all_ entries - not just those of the peer `GetRequestable` was called for. This is necessary, as otherwise you can't fairly decide what to give to which peer (the peer you're calling for may have an announcement that just became ready, but another peer may also have an announcement for the same tx with a lower priority). So it's conceptually true what you say, but only if you think of a full cycle of GetRequestable/RequestedTx for all peers.

I *really* want to avoid having TxRequestTracker depend on being able to call `AlreadyHave` - it would make it much less modularized and harder to test the data structure itself.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+
+    //! The current sequence number. Increases for every announcement. This is used to sort txids returned by
+    //! GetRequestable in announcement order.
+    uint64_t m_sequence{0};
```
That's it, exactly.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+ * above. The same transaction is never re-requested from the same peer, unless the transaction was forgotten about
+ * in the mean time. This happens whenever no candidates remain, or when a transaction is successfully received.
+ *
+ *
```
I added a lot of comments explaining the rationale and some analysis to txrequest.h in #19989 

> "Transaction downloads never guarantee to succeed, namely receiving transaction if you hear about its identifier. Success should be define as a best-effort, guaranteeing we advance towards entry removal for any peer who announced a txid"

I have difficulty parsing what you're trying to say here, but I've added some related comments. Please have a look to see if it's better addressed now.

> "To optimize bandwidth-saving, we limit parallel-fetching and favor sequential fetching (there MUST be only one REQUESTED entry at anytime for any given txid)"

Good idea, done.

> "We don't guarantee fetching again a transaction for which validity/standardness may have change between re-announcements, until next block announcement (a AlreadyHave() triggers all txid entries removal)"

I don't understand this. New block announcements don't affect TxRequestTracker (except for transactions that are in the block, and we don't need anymore).

> "Assuming a fast-inbound attacker-controlled peer and low outbound not-controlled ones, a transaction propagation can be delayed at most for one GETDATA_TX_INTERVAL period"

I added a much more accurate formula.

ariard replied:
> I have difficulty parsing what you're trying to say here, but I've added some related comments. Please have a look to see if it's better addressed now.

IIRC I meaned the following. "Learning about a transaction identifier doesn't guarantee the node will successfully receive the transaction, no more take special steps to ensure it as sending non-solicited GETDATAs. An entry is guarantee to never stuck, assuming the local clock is moving forward"

The first point is a minor. The second underscores an important property of new TxRequestTracker design, its state machine always progress towards a terminal state and should never halt. Though it's so obvious it's worthy to comment.

> I don't understand this. New block announcements don't affect TxRequestTracker (except for transactions that are in the block, and we don't need anymore).

IIRC I meaned the following scenario:
* a peer announces and sends Tx 1, it is rejected due to its unconfirmed parent being invalid
* all other pending requests for Tx 1 are deleted (ForgetTxHash)
* the unconfirmed parent is re-announced, sent and accepted under a different wtxid
* a peer announces Tx 1, the TxRequestTracker will never learn about it due to AlreadyHave()==true
* a block is received, the rejection filter is received
* a peer announces Tx 1, the TxRequestTracker learns and requests it

That said, not reevaluating validity of child transactions when the parent one has changed is already an existing issue. Maybe RequestTx comment could be updated to underscore that a received INV isn't unconditionally process by TxRequestTracker but depends of existing transaction relay state.

sipa replied:
> Learning about a transaction identifier doesn't guarantee the node will successfully receive the transaction, no more take special steps to ensure it as sending non-solicited GETDATAs.

I still don't understand what you're trying to say.

> The second underscores an important property of new TxRequestTracker design, its state machine always progress towards a terminal state and should never halt.

I just today realized that there is some nuance here. As long as you have at least one non-attacker peer that announces a transaction, I believe it is probabilistically true (over time the probability of not reaching a final state, where the transaction is forgotten about, tends to 0), but not more than that. An attacker can disconnect and reconnect, and with (arbitrary) amounts of luck, an attacker can be chosen every time, making the transaction stay arbitrarily long.

I'm not sure it's worth going that deep in comments.

> That said, not reevaluating validity of child transactions when the parent one has changed is already an existing issue. Maybe RequestTx comment could be updated to underscore that a received INV isn't unconditionally process by TxRequestTracker but depends of existing transaction relay state.

I don't think this has anything to do with TxRequestTracker - it just does as it's told, and the current code in net_processing, will not immediately retry for entries in recentRejects. We can document this, but it's a far more high level thing, not specifying the behavior of the TxRequestTracker. Maybe more something for the wiki?

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+ *
+ * - DeletedPeer(peer) deletes all entries for a given peer. It should be called when a peer goes offline.
+ *
+ * - AlreadyHaveTx(txid) deletes all entries for a given txid. It should be called when a transaction is successfully
```
Renamed to `ForgetTx`.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+        >
+    >;
+
+    //! This tracker's main data structure.
```
Added a comment explaining memory usage bounds.

sipa commented on src/txrequest.h:
```diff
@@ -0,0 +1,347 @@
+
+    //! Convert a CANDIDATE_DELAYED entry into a CANDIDATE_READY. If this makes it the new best CANDIDATE_READY (and no
+    //! REQUESTED exists) and better than the CANDIDATE_BEST (if any), it becomes the new CANDIDATE_BEST.
+    void PromoteCandidateNew(typename Index::index<ByTxid>::type::iterator it);
```
Done.
